File: cord_engine/frameworks/langchain.py
"""CORD Framework Adapter — LangChain (Python).

Provides two integration patterns:

1. CORDCallbackHandler — LangChain callback that gates every
   LLM and tool call through CORD. Attach to any chain/agent.

2. wrap_langchain_llm() — Monkey-patches invoke() on an LLM
   instance to run CORD before every call.

Usage (callback):
    from cord_engine.frameworks import CORDCallbackHandler

    handler = CORDCallbackHandler(session_intent="Build a dashboard")
    chain.invoke(input, config={"callbacks": [handler]})

Usage (wrapper):
    from cord_engine.frameworks import wrap_langchain_llm

    llm = wrap_langchain_llm(ChatOpenAI(), session_intent="Build a dashboard")
    llm.invoke("Hello")  # CORD gated
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional, Sequence, Union

from cord_engine import evaluate, Proposal, Decision


class CORDCallbackHandler:
    """LangChain callback handler that gates every LLM/tool call through CORD.

    Raises RuntimeError on BLOCK decisions. Compatible with LangChain's
    BaseCallbackHandler interface (duck-typed to avoid hard dependency).
    """

    def __init__(
        self,
        session_intent: str = "",
        throw_on_block: bool = True,
    ) -> None:
        self.session_intent = session_intent
        self.throw_on_block = throw_on_block

    def on_llm_start(
        self,
        serialized: Dict[str, Any],
        prompts: List[str],
        **kwargs: Any,
    ) -> None:
        """Gate LLM calls through CORD."""
        text = "\n".join(prompts)
        verdict = evaluate(Proposal(
            text=text,
            session_intent=self.session_intent,
        ))
        if verdict.decision == Decision.BLOCK and self.throw_on_block:
            raise RuntimeError(
                f"CORD BLOCK: {', '.join(verdict.reasons)}"
            )

    def on_chat_model_start(
        self,
        serialized: Dict[str, Any],
        messages: List[List[Any]],
        **kwargs: Any,
    ) -> None:
        """Gate chat model calls through CORD."""
        parts: list[str] = []
        for msg_list in messages:
            for msg in msg_list:
                content = getattr(msg, "content", str(msg))
                parts.append(content if isinstance(content, str) else str(content))
        text = "\n".join(parts)
        verdict = evaluate(Proposal(
            text=text,
            session_intent=self.session_intent,
        ))
        if verdict.decision == Decision.BLOCK and self.throw_on_block:
            raise RuntimeError(
                f"CORD BLOCK: {', '.join(verdict.reasons)}"
            )

    def on_tool_start(
        self,
        serialized: Dict[str, Any],
        input_str: str,
        **kwargs: Any,
    ) -> None:
        """Gate tool calls through CORD."""
        tool_name = serialized.get("name", "unknown")
        verdict = evaluate(Proposal(
            text=input_str,
            session_intent=self.session_intent,
        ))
        if verdict.decision == Decision.BLOCK and self.throw_on_block:
            raise RuntimeError(
                f"CORD BLOCK on tool {tool_name}: {', '.join(verdict.reasons)}"
            )

    # LangChain expects these to exist even if they're no-ops
    def on_llm_end(self, response: Any, **kwargs: Any) -> None:
        pass

    def on_llm_error(self, error: BaseException, **kwargs: Any) -> None:
        pass

    def on_tool_end(self, output: str, **kwargs: Any) -> None:
        pass

    def on_tool_error(self, error: BaseException, **kwargs: Any) -> None:
        pass

    def on_chain_start(
        self,
        serialized: Dict[str, Any],
        inputs: Dict[str, Any],
        **kwargs: Any,
    ) -> None:
        pass

    def on_chain_end(self, outputs: Dict[str, Any], **kwargs: Any) -> None:
        pass

    def on_chain_error(self, error: BaseException, **kwargs: Any) -> None:
        pass


def wrap_langchain_llm(llm: Any, session_intent: str = "") -> Any:
    """Wrap a LangChain LLM with CORD enforcement.

    Monkey-patches invoke() so every call passes through CORD first.

    Args:
        llm: LangChain LLM instance (ChatOpenAI, ChatAnthropic, etc.)
        session_intent: Declared session goal for intent drift detection.

    Returns:
        The same LLM instance with CORD-gated invoke().
    """
    original_invoke = llm.invoke

    def guarded_invoke(
        input: Any,
        config: Any = None,
        **kwargs: Any,
    ) -> Any:
        text = input if isinstance(input, str) else str(input)
        verdict = evaluate(Proposal(
            text=text,
            session_intent=session_intent,
        ))
        if verdict.decision == Decision.BLOCK:
            raise RuntimeError(
                f"CORD BLOCK: {', '.join(verdict.reasons)}"
            )
        return original_invoke(input, config=config, **kwargs)

    llm.invoke = guarded_invoke
    return llm

File: cord/frameworks/langchain.js
/**
 * CORD Framework Adapter — LangChain (JavaScript)
 *
 * Wraps LangChain models, chains, and tools with CORD enforcement.
 * Every invoke() call is gated through evaluateProposal() before execution.
 *
 * Usage:
 *   const { wrapLangChain, wrapChain, wrapTool } = require("cord-engine").frameworks;
 *   const model = wrapLangChain(new ChatOpenAI(), { sessionIntent: "Build a dashboard" });
 *   const chain = wrapChain(myChain);
 *   const tool  = wrapTool(myTool);
 */

const { evaluateProposal } = require("../cord");

/**
 * Extract text from LangChain-style input (string, BaseMessage, or object).
 */
function extractText(input) {
  if (typeof input === "string") return input;
  if (input && typeof input.content === "string") return input.content;
  if (Array.isArray(input)) {
    return input.map((m) => (typeof m === "string" ? m : m.content || "")).join("\n");
  }
  return JSON.stringify(input);
}

/**
 * Run CORD evaluation and throw on BLOCK.
 */
function cordGate(text, options = {}) {
  const result = evaluateProposal({
    text,
    toolName: options.toolName || "",
    sessionIntent: options.sessionIntent || "",
    actionType: options.actionType || "",
    useVigil: options.useVigil !== false,
  });

  if (result.decision === "BLOCK") {
    const err = new Error(`CORD BLOCK: ${result.reasons.join(", ")}`);
    err.cordResult = result;
    throw err;
  }
  return result;
}

/**
 * Wrap a LangChain LLM/ChatModel with CORD enforcement.
 * Intercepts invoke(), call(), and generate() methods.
 *
 * @param {object} model    - LangChain model instance (ChatOpenAI, ChatAnthropic, etc.)
 * @param {object} [options] - { sessionIntent, toolName, actionType }
 * @returns {object}         - Proxied model with CORD gates
 */
function wrapLangChain(model, options = {}) {
  const proxy = Object.create(model);

  if (typeof model.invoke === "function") {
    const originalInvoke = model.invoke.bind(model);
    proxy.invoke = async (input, config) => {
      const text = extractText(input);
      cordGate(text, options);
      return originalInvoke(input, config);
    };
  }

  if (typeof model.call === "function") {
    const originalCall = model.call.bind(model);
    proxy.call = async (messages, ...rest) => {
      const text = extractText(messages);
      cordGate(text, options);
      return originalCall(messages, ...rest);
    };
  }

  if (typeof model.generate === "function") {
    const originalGenerate = model.generate.bind(model);
    proxy.generate = async (messages, ...rest) => {
      const text = (Array.isArray(messages) ? messages : [messages])
        .map(extractText).join("\n");
      cordGate(text, options);
      return originalGenerate(messages, ...rest);
    };
  }

  return proxy;
}

/**
 * Wrap a LangChain Chain (RunnableSequence, etc.) with CORD enforcement.
 *
 * @param {object} chain     - LangChain chain instance
 * @param {object} [options] - CORD options
 * @returns {object}         - Proxied chain
 */
function wrapChain(chain, options = {}) {
  const proxy = Object.create(chain);

  if (typeof chain.invoke === "function") {
    const originalInvoke = chain.invoke.bind(chain);
    proxy.invoke = async (input, config) => {
      const text = extractText(input);
      cordGate(text, options);
      return originalInvoke(input, config);
    };
  }

  if (typeof chain.batch === "function") {
    const originalBatch = chain.batch.bind(chain);
    proxy.batch = async (inputs, config) => {
      for (const input of inputs) {
        cordGate(extractText(input), options);
      }
      return originalBatch(inputs, config);
    };
  }

  return proxy;
}

/**
 * Wrap a LangChain Tool with CORD enforcement.
 *
 * @param {object} tool      - LangChain tool instance
 * @param {object} [options] - CORD options (toolName defaults to tool.name)
 * @returns {object}         - Proxied tool
 */
function wrapTool(tool, options = {}) {
  const proxy = Object.create(tool);
  const toolOpts = { ...options, toolName: options.toolName || tool.name || "unknown" };

  if (typeof tool.invoke === "function") {
    const originalInvoke = tool.invoke.bind(tool);
    proxy.invoke = async (input) => {
      const text = extractText(input);
      cordGate(text, toolOpts);
      return originalInvoke(input);
    };
  }

  if (typeof tool.call === "function") {
    const originalCall = tool.call.bind(tool);
    proxy.call = async (input) => {
      const text = extractText(input);
      cordGate(text, toolOpts);
      return originalCall(input);
    };
  }

  return proxy;
}

module.exports = { wrapLangChain, wrapChain, wrapTool };
